src/tracks/playabletrack/wavetrack/ui/SpectrumCache.cpp

```diff
@@ -108,18 +103,19 @@ bool SpecCache::CalculateOneSpectrum
 
    sampleCount from;
 
+   const auto offset = clip.GetSequenceStartTime();
+   const auto numSamples = clip.GetSequenceSamplesCount() / clip.GetWidth();
```

Comment:
Maybe the first real bug I have found:

While dividing by `GetWidth` may correct a mistake introduced earlier -- yet I look at the uses of `SpecCache::Populate` and I see one (in DrawClipSpectrum) passed the visible sample count, while the other (in this file) passed the sequence sample count.  But now both use the sequence count.

Why is this right?

Reply:
Thanks for this question, it motivated me to (try and) understand this whole code better.

I think it is okay the way it is. `numSamples` will be used to know whether the time-domain input to the spectrogram needs zero-padding, if the window goes beyond that. Providing the sequence sample count, as opposed to the _visible_ sample count, allows the code to fetch samples beyond the visible boundary to draw the spectrum, which is good.

The other call used to use `GetVisibleSampleCount`, and leading to unnecessary zero-padding in case of right trimming. Note, though, that that other call only gets executed if the fisheye state is `PINNED`, but it never is, as it's hard-coded to `HIDDEN`, so I cannot test that it indeed works with these changes.

What I could test, though, is that an impulse, whether at the beginning or end of the clip, has its spectrogram displayed properly, also with left- and/or right-trimming.

Reply:
I'll let your decisions stand on this, but maybe it should be reexamined later.  Maybe a clip with unusually large hidden data will cause caching to do too much work and even defeat the purpose of caching for performance.  Maybe there should be a smarter way to bound the cache size, more than the visible samples but fewer than the whole.